photon-lib/src/main/java/org/photonvision/VisionSystemSim.java

```diff
@@ -0,0 +1,369 @@
+    /**
+     * Adds targets on the field which your vision system is designed to detect. The
+     * {@link PhotonCamera}s simulated from this system will report the location of the camera
+     * relative to the subset of these targets which are visible from the given camera position. 
+     * 
+     * @param type Type of target (e.g. "cargo").
+     * @param targets Targets to add to the simulated field
+     */
+    public void addVisionTargets(String type, VisionTargetSim... targets) {
+        if(targetSets.get(type) == null) targetSets.put(type, new HashSet<>());
+        for(var tgt : targets) {
+            targetSets.get(type).add(tgt);
+        }
+    }
+    public void clearVisionTargets() {
+        targetSets.clear();
+    }
+    public Set<VisionTargetSim> removeVisionTargets(String type) {
+        return targetSets.remove(type);
+    }
+    public Set<VisionTargetSim> removeVisionTargets(VisionTargetSim... targets) {
+        var removeList = List.of(targets);
+        var removedSet = new HashSet<VisionTargetSim>();
+        for(var entry : targetSets.entrySet()) {
+            entry.getValue().removeIf(t -> {
+                if(removeList.contains(t)) {
+                    removedSet.add(t);
+                    return true;
+                }
+                else return false;
+            });
+        }
+        return removedSet;
+    }
+
+    /**
+     * Get the latest robot pose in meters saved by the vision system.
+     */
+    public Pose3d getRobotPose() {
+        return getRobotPose(Timer.getFPGATimestamp());
+    }
+    /**
+     * Get the robot pose in meters saved by the vision system at this timestamp.
+     * @param timestamp Timestamp of the desired robot pose
+     */
+    public Pose3d getRobotPose(double timestamp) {
+        return robotPoseBuffer.getSample(timestamp).orElse(new Pose3d());
+    }
+    /**
+     * Clears all previous robot poses and sets robotPose at current time.
+     */
+    public void resetRobotPose(Pose2d robotPose) {
+        resetRobotPose(new Pose3d(robotPose));
+    }
+    /**
+     * Clears all previous robot poses and sets robotPose at current time.
+     */
+    public void resetRobotPose(Pose3d robotPose) {
+        robotPoseBuffer.clear();
+        robotPoseBuffer.addSample(Timer.getFPGATimestamp(), robotPose);
+    }
+
+    public Field2d getDebugField() {
+        return dbgField;
+    }
+
+    /**
+     * Periodic update. Ensure this is called repeatedly-- camera performance is used to
+     * automatically determine if a new frame should be submitted.
+     * 
+     * @param robotPoseMeters The current robot pose in meters
+     */
+    public void update(Pose2d robotPoseMeters) {
+        update(new Pose3d(robotPoseMeters));
+    }
+    /**
+     * Periodic update. Ensure this is called repeatedly-- camera performance is used to
+     * automatically determine if a new frame should be submitted.
+     * 
+     * @param robotPoseMeters The current robot pose in meters

```

Comment:
Should this pose come from my drivetrain simulator?

Reply:
Yes, this is the "actual" simulated pose of your drivetrain on the field. If you want to simulate vision correction, you'll need to separate your noisy/estimated drivetrain pose used for control from an "actual"/noiseless odometry pose.